cmd/otel-allocator/main.go

```diff
@@ -16,306 +16,205 @@ package main
 	}, []string{"source"})
 )
 
 func main() {
+	var (
+		// allocatorPrehook will be nil if filterStrategy is not set or
+		// unrecognized. No filtering will be used in this case.
+		allocatorPrehook prehook.Hook
+		allocator        allocation.Allocator
+		discoveryManager *discovery.Manager
+		collectorWatcher *collector.Client
+		fileWatcher      allocatorWatcher.Watcher
+		promWatcher      allocatorWatcher.Watcher
+		targetDiscoverer *target.Discoverer
+
+		events          chan allocatorWatcher.Event
+		errors          chan error
+		discoveryCancel context.CancelFunc
+		runGroup        run.Group
+	)
 	cliConf, err := config.ParseCLI()
 	if err != nil {
 		setupLog.Error(err, "Failed to parse parameters")
 		os.Exit(1)
 	}
-	cfg, err := config.Load(*cliConf.ConfigFilePath)
-	if err != nil {
-		setupLog.Error(err, "Unable to load configuration")
+	cfg, configLoadErr := config.Load(*cliConf.ConfigFilePath)
+	if configLoadErr != nil {
+		setupLog.Error(configLoadErr, "Unable to load configuration")
 	}
 
 	cliConf.RootLogger.Info("Starting the Target Allocator")
-
 	ctx := context.Background()
-
 	log := ctrl.Log.WithName("allocator")
 
-	// allocatorPrehook will be nil if filterStrategy is not set or
-	// unrecognized. No filtering will be used in this case.
-	allocatorPrehook := prehook.New(cfg.GetTargetsFilterStrategy(), log)
-
-	allocator, err := allocation.New(cfg.GetAllocationStrategy(), log, allocation.WithFilter(allocatorPrehook))
+	events = make(chan allocatorWatcher.Event)
+	errors = make(chan error)
+	allocatorPrehook = prehook.New(cfg.GetTargetsFilterStrategy(), log)
+	allocator, err = allocation.New(cfg.GetAllocationStrategy(), log, allocation.WithFilter(allocatorPrehook))
 	if err != nil {
 		setupLog.Error(err, "Unable to initialize allocation strategy")
 		os.Exit(1)
 	}
-
-	watcher, err := allocatorWatcher.NewWatcher(setupLog, cfg, cliConf, allocator)
-	if err != nil {
-		setupLog.Error(err, "Can't start the watchers")
+	discoveryCtx, discoveryCancel := context.WithCancel(ctx)
+	discoveryManager = discovery.NewManager(discoveryCtx, gokitlog.NewNopLogger())
+	targetDiscoverer = target.NewDiscoverer(log, discoveryManager, allocatorPrehook)
+	collectorWatcher, collectorWatcherErr := collector.NewClient(log, cliConf.ClusterConfig)
+	if collectorWatcherErr != nil {
+		setupLog.Error(collectorWatcherErr, "Unable to initialize collector watcher")
 		os.Exit(1)
 	}
-	defer func() {
-		watcherErr := watcher.Close()
-		if watcherErr != nil {
-			log.Error(watcherErr, "failed to close watcher")
-		}
-	}()
-
-	// creates a new discovery manager
-	discoveryManager := lbdiscovery.NewManager(ctx, log, gokitlog.NewNopLogger(), allocatorPrehook)
-	defer discoveryManager.Close()
-
-	discoveryManager.Watch(allocator.SetTargets)
-
-	k8sclient, err := configureFileDiscovery(context.Background(), log, allocator, discoveryManager, cliConf)
+	fileWatcher, err = allocatorWatcher.NewFileWatcher(setupLog.WithName("file-watcher"), cliConf)
```

Comment:
I think this makes sense, but I would say this deserves its own PR rather than adding it in here. I think this would make even more changes than I already am doing, and I think also needs more discussion too.